### PR TITLE
Implement jemalloc extent hooks memory tracker

### PR DESCRIPTION
**Pull request by antoniofilipovic**
**Idea:**
This PR wants to introduce a memory tracker on top of `jemalloc` instead of using a memory tracker around new and delete. One problem with using a memory tracker on new and delete is that they are not tracking memory allocations happening in the background of jemalloc.

Currently, we are confident that the `jemalloc` tracker is tracking every memory allocation. Test included setting bytes on every page `jemalloc` returned. In that case, RES size was 1:1 with the jemalloc tracker.

 
Problems with current `v4.sh`:
- percpu arena is not obeyed
- oversize threshold is not obeyed
- memory retain is making impossible to track allocations due to lot of virtual memory allocations

New setup script for `jemalloc`:
```
./autogen.sh
MALLOC_CONF="retain:false,percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:5000,dirty_decay_ms:5000" \
./configure \
  --disable-cxx \
  --enable-shared=no --prefix=$working_dir \
  --with-malloc-conf="retain:false,percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:5000,dirty_decay_ms:5000"

```

PR steps to make it work fully:
- [x] add our hooks as a wrapper around `jemalloc` default hooks
- [x] fix issue inside jemalloc with percpu arena not being obeyed
- [x] disable jemalloc oversize_threashold which is used for custom arena on which is not possible to set extant hooks
- [x] fix the issue with the commit hook not being called, and alloc hook only getting calls for virtual memory 
**Note**: this is not fixed in the end, this is jemalloc internal behavior, but commit hook is not called
- [x] add counters to track `alloc`, `dalloc`, `destroy`, `commit`, `decommit`, `purge_forced`, `purge_lazy`, `split`, `merge`
- [x] add check if allocations are multiple of page size 
- [x] fix issue on jemalloc memory tracker increasing memory more than expected:
     explanation: jemalloc calls alloc hook more than necessary, this can be fixed but at a cost. If tcache is disabled, then it means that everything will be backed by allocation and in that case resident memory (RES) will increase a lot on normal workflows
- [x] check if when committing we should reduce size from virtual_allocated, increase on allocated
- [x] check if alignment needs to be included in memory allocation
- [x] swap current memory tracker with `jemalloc` extent hook memory tracker
- [x] ULTIMATE TEST: add bigger size for deltas initial allocation (1MB) and check if benchmark will pass: it did :heavy_check_mark: The thing is now we are tracking memory correctly, and if you allocate that much, there is bug in code
- [x] Check if transparent huge pages in jemalloc have any influence
- [x] check if caching when has any influence
- [x] update libs to have new jemalloc version
- [x] add custom tests to check memory allocation without using the RSS memory tracker behind - the problem is that inside RSS we can have other stuff, like loaded modules, etc

TODO(@antoniofilipovic): Don't forget to check if we need to update `vm.max_map_count`

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Provide the full content or a guide for the final git message

To keep docs changelog up to date, one more thing to do:
- [ ] Write a release note here, including added/changed clauses
- [ ] Tag someone from docs team in the comments

### CODE REVIEW

**Comment by antoniofilipovic** [comment]
There is an additional config that we should check if we want to change:
from `muzzy_decay_ms:5000,dirty_decay_ms:5000` to `muzzy_decay_ms:0,dirty_decay_ms:0`

This will force `jemalloc` to purge memory pages once they are ready to be unused. Otherwise, jemalloc purges pages lazily, once there is a need for them in the system.

File: src/memory/memory_control.cpp [hunk 1]
```diff
@@ -10,18 +10,297 @@
...
+static constexpr extent_hooks_t custom_hooks = {
+    .alloc = &my_alloc,
+    .dalloc = &my_dalloc,
+    .destroy = &my_destroy,
+    .commit = &my_commit,
+    .decommit = &my_decommit,
+    .purge_lazy = &my_purge_lazy,
+    .purge_forced = &my_purge_forced,
+    .split = &my_split,
+    .merge = &my_merge,
```
Comment (Ignition) [inline src/memory/memory_control.cpp:202]: Why not just use old_hooks directly...why add runtime assert + indirection

File: src/memory/new_delete.cpp [hunk 2]
```diff
@@ -41,9 +41,7 @@ void *newImpl(const std::size_t size, const std::align_val_t align) {
 }
 
 void *newNoExcept(const std::size_t size) noexcept { return malloc(size); }
-void *newNoExcept(const std::size_t size, const std::align_val_t align) noexcept {
-  return aligned_alloc(size, static_cast<std::size_t>(align));
-}
+void *newNoExcept(const std::size_t size, const std::align_val_t align) noexcept { return malloc(size); }
```
Comment (Ignition) [inline src/memory/new_delete.cpp:44]: DANGER: why `aligned_alloc` -> `malloc`

File: tests/e2e/memgraph.py [hunk 3]
```diff
@@ -130,9 +130,14 @@ def is_running(self):
     def stop(self):
         if not self.is_running():
             return
-        self.proc_mg.terminate()
-        code = self.proc_mg.wait()
-        assert code == 0, "The Memgraph process exited with non-zero!"
+
+        pid = self.proc_mg.pid
+        try:
+            os.kill(pid, 15)  # 15 is the signal number for SIGTERM
```
Comment (Ignition) [inline tests/e2e/memgraph.py:136]: Why this change?
Popen.terminate() - Stop the child. On POSIX OSs the method sends SIGTERM to the child.

File: src/memory/new_delete.cpp (hunk 2 above)
Comment (antoniofilipovic) [reply src/memory/new_delete.cpp:44]: yes yes, reverted that, it was just some mine testing

File: src/memory/memory_control.cpp (hunk 1 above)
Comment (antoniofilipovic) [reply src/memory/memory_control.cpp:202]: Hmm what do you mean?

Comment (antoniofilipovic) [reply src/memory/memory_control.cpp:202]: I need to use different hooks, because if I use old hooks then jemalloc won't call my hooks and I can't count memory.

File: tests/e2e/memgraph.py (hunk 3 above)
Comment (antoniofilipovic) [reply tests/e2e/memgraph.py:136]: Yes, but there was some issue with this method. It reported that the memgraph didn't properly close, although from logs and by checking manually everything was fine. I checked for different ways to close the memgraph, and this one works. I couldn't see any issue, so I presumed the method was the problem.

File: src/memory/memory_control.cpp (hunk 1 above)
Comment (Ignition) [reply src/memory/memory_control.cpp:202]: But these hooks don't do any counting, they assert + pass to old hooks

Comment (antoniofilipovic) [reply src/memory/memory_control.cpp:218]: Hmm, they do. If you check alloc, purge_forced, dalloc, there are calls to memory tracker?

Comment (antoniofilipovic) [reply src/memory/memory_control.cpp:218]: Ah, you are thinking on only few of them, like merge and split, not purge_forced also. Yes, that makes sense

File: src/query/interpreter.cpp [hunk 4]
```diff
@@ -3068,6 +3067,8 @@ PreparedQuery PrepareSystemInfoQuery(ParsedQuery parsed_query, bool in_explicit_
             {TypedValue("average_degree"), TypedValue(info.average_degree)},
             {TypedValue("memory_usage"), TypedValue(static_cast<int64_t>(info.memory_usage))},
             {TypedValue("disk_usage"), TypedValue(static_cast<int64_t>(info.disk_usage))},
+            // {TypedValue("jemalloc_memory_allocated"),
+            //  TypedValue(utils::GetReadableSize(static_cast<double>(utils::total_memory_tracker.Amount())))},
```
Comment (antoniofilipovic) [inline src/query/interpreter.cpp:3071]: Will move this to other PR or maybe leave it here, as it seems useful for end user, what are opinions on this one?

**Comment by gitbuda** [comment]
STATUS: Ready to be merged

**Review by Ignition — APPROVED** [review]

File: src/memory/memory_control.cpp [hunk 5]
```diff
@@ -22,6 +24,181 @@ namespace memgraph::memory {
...
+
+  sz = sizeof(unsigned);
+  unsigned n_arenas{0};
+  int err = mallctl("opt.narenas", (void *)&n_arenas, &sz, nullptr, 0);
+
+  if (err) {
+    return;
+  }
+
+  spdlog::trace("n areanas {}", n_arenas);
```
Comment (Josipmrden) [inline src/memory/memory_control.cpp:151]: I think we don't need this in the logs

**Review by Josipmrden — COMMENTED** [review]
minor comments, can merge without re-assigning me to review

**Comment by antoniofilipovic** [comment]
@vpavicic RN: Fix problems with memory limit not being honored.

**Comment by antoniofilipovic** [comment]
PR fixes following issue: https://github.com/memgraph/memgraph/issues/1222

**Review by Josipmrden — APPROVED** [review]

**Comment by antoniofilipovic** [comment]
@vpavicic I need to add docs page for VM max map count, similar to this one: https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#_set_vm_max_map_count_to_at_least_262144
